packages/nextjs/src/client/routing/appRouterRoutingInstrumentation.ts

```diff
@@ -21,70 +23,111 @@ export function appRouterInstrumentPageLoad(client: Client): void {
+  const checkForRouterAvailabilityInterval = setInterval(() => {
+    triesToFindRouter++;
+    const router = GLOBAL_OBJ_WITH_NEXT_ROUTER?.next?.router ?? GLOBAL_OBJ_WITH_NEXT_ROUTER?.nd?.router;
 
-  // Make sure the second argument is some kind of fetch config obj that contains headers
-  if (!fetchArgs[1] || typeof fetchArgs[1] !== 'object' || !('headers' in fetchArgs[1])) {
-    return null;
-  }
+    if (routerPatched || triesToFindRouter > MAX_TRIES_TO_FIND_ROUTER) {
+      clearInterval(checkForRouterAvailabilityInterval);
+    } else if (router) {
+      clearInterval(checkForRouterAvailabilityInterval);
+      routerPatched = true;
+      (['back', 'forward', 'push', 'replace'] as const).forEach(routerFunctionName => {
+        if (router?.[routerFunctionName]) {
+          // @ts-expect-error Weird type error related to not knowing how to associate return values with the individual functions - we can just ignore
+          router[routerFunctionName] = new Proxy(router[routerFunctionName], {
+            apply(target, thisArg, argArray) {
+              const span = startBrowserTracingNavigationSpan(client, {
+                name: INCOMPLETE_APP_ROUTER_INSTRUMENTATION_TRANSACTION_NAME,
+                attributes: {
+                  [SEMANTIC_ATTRIBUTE_SENTRY_OP]: 'navigation',
+                  [SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: 'auto.navigation.nextjs.app_router_instrumentation',
+                  [SEMANTIC_ATTRIBUTE_SENTRY_SOURCE]: 'url',
+                },
+              });
 
-  try {
-    const url = fetchArgs[0] as URL;
-    const headers = fetchArgs[1].headers as Record<string, string>;
+              currentNavigationSpan = span;
 
-    // Not an RSC request
-    if (headers['RSC'] !== '1') {
-      return null;
-    }
+              if (routerFunctionName === 'push') {
+                span?.updateName(transactionNameifyRouterArgument(argArray[0]));
+                span?.setAttribute('navigation.type', 'router.push');
+              } else if (routerFunctionName === 'replace') {
+                span?.updateName(transactionNameifyRouterArgument(argArray[0]));
+                span?.setAttribute('navigation.type', 'router.replace');
+              } else if (routerFunctionName === 'back') {
+                span?.setAttribute('navigation.type', 'router.back');
+              } else if (routerFunctionName === 'forward') {
+                span?.setAttribute('navigation.type', 'router.forward');
+              }
 
-    // Prefetch requests are not navigating RSC requests
-    if (headers['Next-Router-Prefetch'] === '1') {
-      return null;
+              return target.apply(thisArg, argArray);
+            },
+          });
+        }
+      });
     }
+  }, ROUTER_AVAILABILITY_CHECK_INTERVAL_MS);
+}
 
-    return {
-      targetPathname: url.pathname,
-    };
+function transactionNameifyRouterArgument(target: string): string {
+  try {
+    return new URL(target, 'http://some-random-base.com/').pathname;
```

Comment:
q: Why do we need that exactly? Couldn't we just take the target?

Reply:
So users could be passing basically any string to `router.push()`, including absolute URLs to the current origin. Taking the absolute URLs wouldn't be good, so we somehow need to parse out the path. Parsing urls is really hard so this is the easiest/best way I found.